fastlane_core/lib/fastlane_core/itunes_transporter.rb

```diff
@@ -180,9 +180,130 @@ def additional_upload_parameters
     end
   end
 
+  # Generates commands and executes the altool.
+  class AltoolTransporterExecutor < TransporterExecutor
+    ERROR_REGEX = /\*\*\* Error:\s+(.+)/
+
+    private_constant :ERROR_REGEX
+
+    def execute(command, hide_output)
+      if Helper.test?
+        yield(nil) if block_given?
+        return command
+      end
+
+      @errors = []
+      @all_lines = []
+
+      if hide_output
+        # Show a one time message instead
+        UI.success("Waiting for App Store Connect transporter to be finished.")
+        UI.success("Application Loader progress... this might take a few minutes...")
+      end
+
+      begin
+        exit_status = FastlaneCore::FastlanePty.spawn(command) do |command_stdout, command_stdin, pid|
+          command_stdout.each do |line|
+            @all_lines << line
+            parse_line(line, hide_output) # this is where the parsing happens
+          end
+        end
+      rescue => ex
+        # FastlanePty adds exit_status on to StandardError so every error will have a status code
+        exit_status = ex.exit_status
+        @errors << ex.to_s
+      end
+
+      @errors << "The call to the altool completed with a non-zero exit status: #{exit_status}. This indicates a failure." unless exit_status.zero?
+
+      unless @errors.empty? || @all_lines.empty?
+        # Print the last lines that appear after the last error from the logs
+        # If error text is not detected, it will be 20 lines
+        # This is key for non-verbose mode
+
+        # The format of altool's result with error is like below
+        # > *** Error: Error uploading '...'.
+        # > *** Error: ...
+        # > {
+        # >     NSLocalizedDescription = "...",
+        # >     ...
+        # > }
+        # So this line tries to find the line which has "*** Error:" prefix from bottom of log
+        error_line_index = @all_lines.rindex { |line| ERROR_REGEX.match?(line) }
+
+        @all_lines[(error_line_index || -20)..-1].each do |line|
+          UI.important("[altool] #{line}")
+        end
+        UI.message("Application Loader output above ^")
+        @errors.each { |error| UI.error(error) }
+      end
+
+      yield(@all_lines) if block_given?
+      exit_status.zero?
+    end
+
+    def build_upload_command(username, password, source = "/tmp", provider_short_name = "", jwt = nil, platform = nil, api_key = nil)
```

Comment:
I'd love for this method to use named parameters at some point (`build_upload_command(username:, password:, source: "/tmp", provider_short_name: "", jwt: nil, platform: nil, api_key: nil)` for easier-to-read call sites; but that's just a code-style nitpick and not a blocker at all, and can be ironed out in a separate PR, so no worries.

Reply:
I see, since `build_upload_command`'s style is same with other executer class so it would be kind of refactoring. I think I can do it later PR 🙇

Reply:
Totally 👍  no need to fix this in this PR given the refactoring it will involve indeed. And not that critical if we don't do it anyway to be honest 🙃